### PR TITLE
refactor(whl_library): move bazel file generation to Starlark

### PR DESCRIPTION
**Pull request by aignas**
Before this PR, the `wheel_installer` was doing three things:
1. Downloading the right wheel.
2. Extracting it into the output directory.
3. Generating BUILD.bazel files based on the extracted contents.

This PR is moving the third part into the `whl_library` repository rule
and it has the following benefits:
* We can reduce code duplication and label sanitization functions in
  rules_python.
* There are many things that the `wheel_installer` does not care anymore
  and we don't need to change less code when extending `whl_library` as
  we can now do many things in starlark directly.
* It becomes easier to change the API of how we expose the generated
  BUILD.bazel patching because we only need to change the Starlark
  functions.

Work towards #1330.

### CODE REVIEW

**Comment by aignas** [comment]
FYI @chrislovecnm, I decided to see how this idea of generating BUILD.bazel and other extra files worked out in reality and I think I am happy with the solution. It makes working on the #1262 and extending the API easier, hence me doing this refactor before extending the `whl_library` more.

File: python/pip_install/private/generate_whl_library_build_bazel.bzl [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    srcs = glob(["*.whl"], allow_empty = True),
+    data = {whl_file_deps},
+)
+
+py_library(
+    name = "{name}",
+    srcs = glob(
+        ["site-packages/**/*.py"],
+        exclude={srcs_exclude},
+        allow_empty = True,
```
Comment (rickeylev) [inline python/pip_install/private/generate_whl_library_build_bazel.bzl:72]:
```suggestion
        # Empty sources are allowed to support wheels that don't have any
        # pure-Python code, e.g. pymssql, which is written in Cython.
        allow_empty = True,
```

File: python/pip_install/private/generate_whl_library_build_bazel.bzl [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        data_exclude: more patterns to exclude from the data attribute of generated py_library rules.
+        tags: list of tags to apply to generated py_library rules.
+        entry_points: A dict of entry points to add py_binary rules for.
+        annotation: The annotation for the build file.
+
+    Returns:
+        A complete BUILD file as a string
+
+    We allow for empty Python sources as for Wheels containing only compiled C code
+    there may be no Python sources whatsoever (e.g. packages written in Cython: like `pymssql`).
```
Comment (rickeylev) [inline python/pip_install/private/generate_whl_library_build_bazel.bzl:107]: nit: move the comment inline where the allow_empty setting is actually used (see the other suggested edit)

```suggestion
```

File: python/pip_install/private/generate_whl_library_build_bazel.bzl [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    additional_content = []
+    data = []
+    srcs_exclude = []
+    data_exclude = [] + data_exclude
+    dependencies = sorted(dependencies)
+    tags = sorted(tags)
+
+    for entry_point, entry_point_script_name in entry_points.items():
+        additional_content.append(
+            _generate_entry_point_rule(
```
Comment (rickeylev) [inline python/pip_install/private/generate_whl_library_build_bazel.bzl:119]: nit: use named args for this call.

File: python/pip_install/private/generate_whl_library_build_bazel.bzl [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        script (str): The path to the entry point's python file.
+        pkg (str): The package owning the entry point. This is expected to
+            match up with the `py_library` defined for each repository.
+
+    Returns:
+        str: A `py_binary` instantiation.
+    """
+    return _ENTRY_POINT_RULE_TEMPLATE.format(
+        name = name,
+        src = str(script).replace("\\", "/"),
```
Comment (rickeylev) [inline python/pip_install/private/generate_whl_library_build_bazel.bzl:223]: This str() call looks unnecessary? The input is already a string, no?

**Review by rickeylev — CHANGES_REQUESTED** [review]
Mostly lgtm

**Review by rickeylev — APPROVED** [review]

**Pull request merged**